### ISSUE TITLE
[Pipeline/Model] PhotoMaker

### ISSUE THREAD

**Issue by a-r-r-o-w** [huggingface/diffusers#6579]
Highlights from the paper:
- Arbitrary number of identity images can be provided to create the "stacked ID embedding" - a representation of the input images that the model is trained to understand. Input images and prompt are encoded and combined to create the stacked embedding, which is fed into the cross attention layers of the unet (implementation detail: this new stacked embedding is the new `prompt_embeds` instead of it being just the actual prompt clip-embedding.)
- Even though multiple images are used to drive the generation, performance impact is minimal.
- Tuning-free i.e. does not require additional finetuning like DreamBooth.
- Proposes a data construction pipeline that to build a human-centric dataset.
- Metrics on par (first/second-best on most metrics) with existing methods.

<table>
<tr>
<td align=center>Realistic</td>
<td align=center>Stylistic</td>
</tr>
<tr>
<td align=center><img src="https://github.com/huggingface/diffusers/assets/72266394/10bb6cf7-890c-4d1d-98f8-e253d3e06448"></td>
<td align=center><img src="https://github.com/huggingface/diffusers/assets/72266394/7c7e5b00-d61f-4837-a464-fe7eb5ecf615"></td>
</tr>
</table>

Code: https://github.com/TencentARC/PhotoMaker
Project site: https://photo-maker.github.io/
Demo: [this](https://huggingface.co/spaces/TencentARC/PhotoMaker) and [this](https://huggingface.co/spaces/TencentARC/PhotoMaker-Style)
Model: https://huggingface.co/TencentARC/PhotoMaker

Authors: @paper99 @ljzycmd @xinntao @zhongangqi @MingMingCheng [Ying Shan]

Thank you for you amazing research, and very intuitive and well-written paper that someone newer to the field can easily understand! Opening the issue in case there is interest to add PhotoMaker adapters to diffusers.

@sayakpaul @patrickvonplaten 

The codebase is built upon diffusers and can be easily integrated.

**Comment by sayakpaul** [comment]
The codebase already leverages `diffusers` fully. So, I would maybe honor the original representation. They also have a demo: https://huggingface.co/spaces/TencentARC. 

Going through the code from that demo, seems to be very easily usable.

**Comment by steverhoades** [comment]
> The codebase already leverages `diffusers` fully. So, I would maybe honor the original representation. They also have a demo: https://huggingface.co/spaces/TencentARC.
> 
> Going through the code from that demo, seems to be very easily usable.

Indeed! Posted some details of my testing [here](https://github.com/TencentARC/PhotoMaker/discussions/14) for those interested. The performance isn't great but the results are pretty good.

**Comment by sayakpaul** [comment]
> The performance isn't great

What do you mean by that? Inference latency?

**Comment by steverhoades** [comment]
> > The performance isn't great
> 
> What do you mean by that? Inference latency?

Meaning that the pipeline execution is fairly slow compared to executing without PhotoMaker. (anecodotal as I have not measured it)

**Comment by sayakpaul** [comment]
That could be probably because you're not following the exact same setup. Some potential suspects could be the use of FP16, SDPA (we default to SDPA when using PyTorch 2.0), etc. 

Interesting that this is the case because PhotoMaker internally uses `diffusers`. See here: https://huggingface.co/spaces/TencentARC/PhotoMaker/tree/main.

**Comment by a-r-r-o-w** [comment]
@steverhoades Would you be interested in taking this up as a PR?

@sayakpaul Maybe you add the contributions-welcome label if someone wants to take this up :)

**Comment by sayakpaul** [comment]
@yiyixuxu here. How do we want to take this up? `research_projects` (as it has new weights).

**Comment by hipsterusername** [comment]
Would definitely be interested in exploring how this performs in practical applications - Seems to have some significant opportunity to drive improved character consistency for creatives.

**Comment by yiyixuxu** [comment]
Like @sayakpaul said, it is already a "diffusers pipeline," - so I'm not sure if there is much to be gained by adding to our community or research folder. I wouldn't mind it though :) 

I guess the question is whether we want to integrate it the same way we integrate IP-Adapter? I think we would need to see more evidence about potential usage before we do that:)

**Comment by hipsterusername** [comment]
> Like @sayakpaul said, it is already a "diffusers pipeline," - so I'm not sure if there is much to be gained by adding to our community or research folder. I wouldn't mind it though :)
> 
> I guess the question is whether we want to integrate it the same way we integrate IP-Adapter? I think we would need to see more evidence about potential usage before we do that:)

Well the pipeline as is is very inefficient, from what I understand. Someone in the community ran the code and said as written running locally is impossible.

A practical implementation seems to be reimplementing in a similar way to IP Adapter, while also optimizing for consumer cards if at all possible.

Recognize that’s a tall order, mainly raising it for discussion

**Comment by yiyixuxu** [comment]
@hipsterusername 
keep us updated on your thoughts on this model:) 
happy to integrate it once it becomes obvious to us that it will have a lot of usage and impact

**Comment by Paper99** [comment]
Hi @a-r-r-o-w @steverhoades @yiyixuxu @sayakpaul @hipsterusername,
Thanks for your efforts.
We are very happy to integrate PhotoMaker into diffusers, if there is anything we can help with, please feel free to contact us.

Besides, there is a strange problem when we use dtype=torch.bfloat16 on a GPU that does not support bfloat16, there will be a significant slowdown. How to avoid this problem? Any good ideas? See this [link](https://github.com/TencentARC/PhotoMaker?tab=readme-ov-file#-new-featuresupdates) for details.

**Comment by sayakpaul** [comment]
I think that is somewhat expected. bfloat16 comoutationa on the GPUs without ampere cores might slow things down.

**Comment by Paper99** [comment]
Thank you for your answer. We have another issue to bring to your attention. Please take a look at this URL: https://github.com/TencentARC/PhotoMaker/issues/98.

**Comment by sayakpaul** [comment]
I think it's okay to be closed since the original implementation already relies on `diffusers`.
